### ISSUE TITLE
CSV generator is not tread-safe

### ISSUE THREAD

**Issue by adityapatadia** [fizyr/keras-retinanet#822]
When I use `workers=8` in keras `fit_generator`, it gives following error:

```
RuntimeError: Your generator is NOT thread-safe.Keras requires a thread-safe generator when`use_multiprocessing=False, workers > 1`.For more information see issue #1638.
```

**Comment by hgaiser** [comment]
Do you happen to know why it is not thread safe?

**Comment by adityapatadia** [comment]
No idea.

**Comment by de-vri-es** [comment]
Native generators in python (the type that are a function using `yield`) can not be called from multiple threads at the same time. Doing so anyway will cause a race that may result in python throwing an exception. To prevent that race, some checks were added to Keras that require a generator class to specifically claim they are thread safe before Keras will call `next()` on them from multiple threads.

However, there is probably little to be gained for the generators that we got thanks to the python global interpreter lock. To bypass the GIL, there's multiprocessing, but that would cause the generator to be copied across different processes and generate the same sequence in each of them.

The best solution is #824. Sequences don't suffer from the same problems that plain generators do in the case of multiprocessing.

**Comment by adityapatadia** [comment]
I don't know whether it's due to generators but the GPU does not get used 100% when training model. It fluctuates. On the other hand, if I train using tensorflow code, it uses GPU 100%.

**Comment by de-vri-es** [comment]
Could you try again with current master? #824 is now merged. It should allow for workers > 1, both with and without multiprocessing. I'd be interested to know if it can bring your GPU utilization up.

If increasing workers alone doesn't help, could you also try with multiprocessing enabled?

**Comment by de-vri-es** [comment]
@adityapatadia: Is this still an issue for you with current master?

**Comment by adityapatadia** [comment]
Haven't tried it. I will try after some days and report if it works.

**Comment by hgaiser** [comment]
@adityapatadia any update on this ?

**Comment by adityapatadia** [comment]
I am sorry I could not try this. I again went back to training with tensorflow as this repo is only limited to retinanet. I don't think I will have time to test anytime soon. Should I close this?

**Comment by hgaiser** [comment]
Yeah let's close this for now then. Thanks for the feedback!

**Comment by ktobah** [comment]
I am having an issue with setting multiprocessing = True. My training freezes and docker exits with code 139. 
If I set it to False, and set workers > 2, same thing happens.
If I set it to False with workers = 2, then my training starts then it stops after a while with an error message saying: "ValueError: generator already executing"
if set it to False and workers = 1, the training runs but GPU usage is very low.

Sometimes, I change the tensorflow-gpu docker image from 1.15.0 to 1.14.0 and vice versa while setting multiprocessing = True and workers = 2, it works (GPU usage around 60% for each GPU). But this is not stable, it works rarely and doesn't work most of the time!

I tried with tf 1.14.0, 1.15.0, and with 2.0.0 but that latter can't usage multi-gpu unless a lot changes (Multi Strategy) to the code are made.
Our training set is around 43000 images and I can use 2 GPUs on our server (has 4).

Any suggestion to why this happening?
